File: agent-service/backend/utils/create_signature.py
import json
import hmac
import hashlib
from datetime import datetime

from backend.config import main as config


def create_hmac_signature(body: dict = {}, query: dict = {}, mode: str = "body"):
    timestamp = int(datetime.timestamp(datetime.utcnow()) * 1000)

    if mode == "body":
        message = json.dumps(body, separators=(",", ":")) + str(timestamp)
    elif mode == "query":
        message = json.dumps(query, separators=(",", ":")) + str(timestamp)
    elif mode == "both":
        message = json.dumps(query, separators=(",", ":")) + json.dumps(body) + str(timestamp)
    else:
        raise ValueError("Invalid mode for signature creation")
    
    message = message.encode('utf-8')
    
    signature = hmac.new(config.HMAC_AGENT_KEY.encode('utf-8'), message, hashlib.sha256).hexdigest()
    return {"signature": signature, "timestamp": timestamp}

File: user-service/utils/createSignature.js
import crypto from 'crypto'
import config from '@/config'

export const createSignature = (body = {}, query = {}, mode = 'body') => {
    let message = ''
    const timestamp = String(Date.now())

    if (mode == 'body') {
        message = JSON.stringify(body) + timestamp
    } else if (mode == 'query') {
        message = JSON.stringify(query) + timestamp
    } else if (mode == 'both') {
        message = JSON.stringify(query) + JSON.stringify(body) + timestamp
    } else {
        throw TypeError('Invalid type mode')
    }

    const signature = crypto.createHmac('SHA256', Buffer.from(config.HMAC_USER_KEY, 'utf-8'))
    signature.update(message)

    return { signature: signature.digest('hex'), timestamp: timestamp }
}